game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/FirstStrikeStepOrder.java

```diff
@@ -69,9 +43,23 @@ private static boolean hasAttackingFirstStrike(final BattleState battleState) {
     return battleState.getAttackingUnits().stream().anyMatch(Matches.unitIsFirstStrike());
   }
 
-  private static boolean hasDefendingFirstStrike(final BattleState battleState) {
-    return battleState.getDefendingUnits().stream()
-        .anyMatch(Matches.unitIsFirstStrikeOnDefense(battleState.getGameData()));
+  private static Optional<FirstStrikeStepOrder> calculateAttackerSteps(
+      final @NonNull BattleState battleState) {
+    final ReturnFire returnFireAgainstAttackingSubs =
+        returnFireAgainstAttackingSubs(
+            battleState.getAttackingUnits(),
+            battleState.getDefendingUnits(),
+            battleState.getGameData());
+    switch (returnFireAgainstAttackingSubs) {
+      case ALL:
+        return Optional.of(OFFENDER_NO_SNEAK_ATTACK);
+      case SUBS:
+        return Optional.of(OFFENDER_SNEAK_ATTACK_WITH_OPPOSING_FIRST_STRIKE);
+      case NONE:
+        return Optional.of(OFFENDER_SNEAK_ATTACK);
+      default:
+        return Optional.empty();
```

Comment:
Do we ever expect to have any other options? If no, then throwing an exception could be valuable in case we fail to update this list.

Have you considered the "polymorphic switch" refactor (I forget the real name for that kind of update), it could potentially remove the switch statement.

EG:

```
returnFireAgainstAttackingSubs.getFirstStrikeStepOrder()
```

Though, I do see we could get a circular dependency. This implies a static lookup method on the enum could do the trick, eg:
```
  public static Optional<FirstStrikeStepOrder> lookupByReturnFireType(ReturnFire returnFire) {
    return Arrays.stream(values())
        .filter(firstStrikeStepOrder -> firstStrikeStepOrder.returnFire == returnFire)
        .findAny();     
  }
```

Yet, I see there are duplicated by offender and attacker. Have you considered splitting the enum into two, one for attacker and another for deferender? In such a case you would not have to encode attacker vs defender in the enum name. Perhaps we could even get the attacker or defender to be a flag value itself so that we do not need as many enum values.

WDYT?